src/supplemental/http/http_server.c

```diff
@@ -1146,26 +1146,31 @@ nni_http_server_add_handler(nni_http_server *s, nni_http_handler *h)
 		while ((len2 > 0) && (h2->uri[len2 - 1] == '/')) {
 			len2--; // ignore trailing '/'
 		}
-		if (strncmp(h->uri, h2->uri, len > len2 ? len2 : len) != 0) {
-			continue; // prefixes don't match.
+
+		if (len != len2) {
+			continue; // length mismatch
 		}
 
-		if (len2 > len) {
-			if ((h2->uri[len] == '/') && (h->tree)) {
-				nni_mtx_unlock(&s->mtx);
-				return (NNG_EADDRINUSE);
-			}
-		} else if (len > len2) {
-			if ((h->uri[len2] == '/') && (h2->tree)) {
-				nni_mtx_unlock(&s->mtx);
-				return (NNG_EADDRINUSE);
+		if (strncmp(h->uri, h2->uri, len) != 0) {
```

Comment:
Given the lengths are the same (above), you can just use strcmp(). here.

Reply:
> Essentially, I think you've missed the point of the h->tree parameter, which is that you have the option of either registering for ownership of the entire tree, or for an exact URL.
> 
> This distinction is really important, because in some use cases (REST APIs) we actually want to put parameters on the URL line -- e.g. GET /object/ -- we would have a handler for GET /object/ and then look at the full path information.
> 

Ok, I see your point. And I now understand why I encountered the "address conflict" problem. In my C++ REST framework (https://github.com/robiwano/siesta) I want to be able to register REST handlers with regular expressions, which means I need to register a tree handler, and resolve the path wrt regexes in my handler callback. So basically, I registered multiple tree handlers which in the current nng implementation induces a conflict. I will address this problem in my framework instead so that if a full path without regexes are given, the handler is registered without the tree option. 

> The idea of having a "fallback" handler, such that the tree logic is used, but we allow a more specific handler to be registered without considering it a conflict, is a good idea.
> 
> I think there is design work required.

Yes, this would be something else.